src/devices/sound/ymadpcm.h

```diff
@@ -0,0 +1,334 @@
+	u8 pan_left() const         /*  1 bit  */ { return BIT(m_regdata[m_chbase + 0x08], 7); }
+	u8 pan_right() const        /*  1 bit  */ { return BIT(m_regdata[m_chbase + 0x08], 6); }
+	u8 instrument_level() const /*  5 bits */ { return BIT(m_regdata[m_chbase + 0x08], 0, 5); }
+	u16 start() const           /* 16 bits */ { return m_regdata[m_chbase + 0x10] | (m_regdata[m_chbase + 0x18] << 8); }
+	u16 end() const             /* 16 bits */ { return m_regdata[m_chbase + 0x20] | (m_regdata[m_chbase + 0x28] << 8); }
+
+	// per-channel writes
+	void write_start(u16 address) { write(m_chbase + 0x10, address); write(m_chbase + 0x18, address >> 8); }
+	void write_end(u16 address) { write(m_chbase + 0x20, address); write(m_chbase + 0x28, address >> 8); }
+
+private:
+	// internal state
+	u8 m_chbase;                   // base offset for channel-specific data
+	std::vector<u8> &m_regdata;    // reference to the raw data
+};
+
+
+// ======================> ymadpcm_a_channel
+
+class ymadpcm_a_channel
+{
+public:
+	// constructor
+	ymadpcm_a_channel(ymadpcm_a_registers regs, read8sm_delegate reader, u8 addrshift);
+
+	// register for save states
+	void save(device_t &device, u8 index);
+
+	// reset the channel state
+	void reset();
+
+	// signal key on/off
+	void keyonoff(bool on);
+
+	// master clockingfunction
+	bool clock();
+
+	// return the computed output value, with panning applied
+	void output(s32 &lsum, s32 &rsum) const;
+
+	// direct parameter setting for YM2608 ROM-based samples
+	void set_start_end(u16 start, u16 end) { m_regs.write_start(start); m_regs.write_end(end); }
+
+private:
+	// internal state
+	u8 const m_address_shift;        // address bits shift-left
+	read8sm_delegate const m_reader; // read delegate
+	u8 m_playing;                    // currently playing?
+	u8 m_curnibble;                  // index of the current nibble
+	u8 m_curbyte;                    // current byte of data
+	u32 m_curaddress;                // current address
+	s16 m_accumulator;               // accumulator
+	s8 m_step_index;                 // index in the stepping table
+	ymadpcm_a_registers m_regs;      // register accessor
+};
+
+
+// ======================> ymadpcm_a_engine
+
+class ymadpcm_a_engine
+{
+public:
+	// constructor
+	ymadpcm_a_engine(device_t &device, read8sm_delegate reader, u8 addrshift);
+
+	// save state handling
+	void save(device_t &device);
+
+	// reset our status
+	void reset();
+
+	// master clocking function
+	u8 clock(u8 chanmask);
+
+	// compute sum of channel outputs
+	void output(s32 &lsum, s32 &rsum, u8 chanmask);
+
+	// write to the ADPCM-A registers
+	void write(u8 regnum, u8 data);
+
+	// set the start/end address for a channel (for hardcoded YM2608 percussion)
+	void set_start_end(u8 chnum, u16 start, u16 end) { m_channel[chnum]->set_start_end(start, end); }
+
+private:
+	// internal state
+	std::vector<std::unique_ptr<ymadpcm_a_channel>> m_channel; // vector of channels
+	std::vector<u8> m_regdata;       // raw register data
```

Comment:
Is there a good reason for using a vector for this?  You could use an array or a `std:array`.  All the vector seems to be doing is hurting locality for a relatively small thing allocated at construction time.  (If you need a reference to an array remember you can use `T (&v)[N]` syntax.)